client/src/_dispatch.rs

```diff
@@ -0,0 +1,290 @@
+use crate::_raw_session::RawSession;
+
+use ya_relay_proto::codec;
+use ya_relay_proto::proto::{self, RequestId};
+
+pub type ErrorHandler = Box<dyn Fn() -> ErrorHandlerResult + Send>;
+pub type ErrorHandlerResult = Pin<Box<dyn Future<Output = ()> + Send>>;
+type ResponseSender = Sender<Dispatched<proto::response::Kind>>;
+
+/// Signals receipt of a response packet
+pub async fn dispatch<H, S>(handler: H, mut stream: S)
+where
+    H: Handler + Clone + 'static,
+    S: Stream<Item = (codec::PacketKind, SocketAddr, chrono::DateTime<chrono::Utc>)> + Unpin,
+{
+    while let Some((packet, from, _timestamp)) = stream.next().await {
+        let handler = handler.clone();
+
+        // First look for existing session, but if it doesn't exist, maybe we
+        // can find dispatcher from temporary session that is being initialized at this moment.
+        let session = handler.session(from).await;
+        let dispatcher = match session.clone() {
+            Some(session) => Some(session.raw.clone()),
+            None => handler.dispatcher(from).await,
+        };
+
+        if let Some(ref dispatcher) = dispatcher {
+            dispatcher.dispatcher.update_seen();
+        }
+
+        match packet {
+            codec::PacketKind::Packet(proto::Packet {
+                session_id,
+                kind: Some(kind),
+            }) => match kind {
+                proto::packet::Kind::Control(control) => {
+                    handler
+                        .on_control(session_id, control, from)
+                        .map(spawn_local);
+                }
+                proto::packet::Kind::Request(request) => {
+                    handler
+                        .on_request(session_id, request, from)
+                        .map(spawn_local);
+                }
+                proto::packet::Kind::Response(response) => {
+                    match response.kind {
+                        Some(kind) => match dispatcher {
+                            Some(dispatcher) => dispatcher.dispatcher.dispatch_response(
+                                from,
+                                response.request_id,
+                                session_id,
+                                response.code,
+                                kind,
+                            ),
+                            None => log::debug!("Unexpected response from {from}: {kind:?}"),
+                        },
+                        // TODO: Handle empty packet kind here
+                        None => log::debug!("Empty response kind from: {from}"),
+                    }
+                }
+            },
+            codec::PacketKind::Forward(forward) => {
+                // In case of temporary sessions we shouldn't get `Forward` packets,
+                // so we can safely ignore this case.
+                if let Some(session) = session {
+                    handler
+                        .on_forward(forward, from, Some(session))
```

Comment:
Maybe `on_forward` should not accept session as an argument, because it can do it by itself inside? I don;t know however what to do with `update_seen`  though

Reply:
This require second lookup and redesign was aiming to minimize number of lookups.